src/isomorphic/classic/element/ReactElement.js

```diff
@@ -158,7 +163,49 @@ ReactElement.createElement = function(type, config, children) {
       }
     }
   }
-
+  if (__DEV__) {
+    // Create dummy `key` and `ref` property to `props` to warn users
+    // against its use
+    if (typeof props.$$typeof === 'undefined' ||
+        props.$$typeof !== REACT_ELEMENT_TYPE) {
+      if (!props.hasOwnProperty('key')) {
+        Object.defineProperty(props, 'key', {
+          get: function() {
+            if (!specialPropKeyWarningShown) {
+              specialPropKeyWarningShown = true;
+              warning(
+                false,
+                '%s: `key` is not a prop. Trying to access it will result ' +
+                  'in `undefined` being returned. If you need to access the same ' +
+                  'value within the child component, you should pass it as a different ' +
+                  'prop. (https://fb.me/react-special-props)',
+                'displayName' in type ? type.displayName: 'Element'
+              );
+            }
```

Comment:
Let's explicitly return undefined here.  This is a getter, so we should return a value, and it makes the expected behavior a little more clear.  Same with the getter for ref.